File: metisfl/controller/core/controller.cc

#include "metisfl/controller/core/controller.h"

namespace metisfl::controller {

// Constructor
Controller::Controller(const GlobalTrainParams &global_train_params,
                       const ModelStoreParams &model_store_params) {
  global_train_params_ = global_train_params;

  learner_manager_ = absl::make_unique<LearnerManager>();
  scheduler_ = CreateScheduler(global_train_params_.scheduler);
  selector_ = CreateSelector();
  model_manager_ =
      absl::make_unique<ModelManager>(learner_manager_.get(), selector_.get(),
                                      global_train_params_, model_store_params);
}

// Public methods
absl::StatusOr<std::string> Controller::AddLearner(const Learner &learner) {
  auto is_semi_sync =
      global_train_params_.scheduler == "SemiSynchronous";
  return learner_manager_->AddLearner(learner, is_semi_sync,
                                      global_train_params_.scaling_factor);
}

absl::Status Controller::RemoveLearner(std::string learner_id) {
  if (!learner_manager_->ValidateLearner(learner_id))
    return absl::NotFoundError("Learner does not exist.");

  std::vector<std::string> learner_ids = {learner_id};
  model_manager_->EraseModels(learner_ids);

  return learner_manager_->RemoveLearner(learner_id);
}

absl::Status Controller::SetInitialModel(const Model &model) {
  return model_manager_->SetInitialModel(model);
}

absl::Status Controller::StartTraining() {
  if (!model_manager_->IsInitialized())
    return absl::FailedPreconditionError("Model is not initialized.");

  auto learner_ids = learner_manager_->GetLearnerIds();
  std::vector<std::string> to_schedule;
  for (const auto &learner_id : learner_ids) {
    to_schedule = scheduler_->ScheduleNext(learner_id, learner_ids.size());
    if (!to_schedule.empty())
      learner_manager_->ScheduleTrain(to_schedule, model_manager_->GetModel());
  }

  return absl::OkStatus();
}

absl::Status Controller::TrainDone(const TrainDoneRequest &request) {
  std::lock_guard<std::mutex> model_manager_guard(model_manager_mutex_);
  std::lock_guard<std::mutex> learner_manager_guard(learner_manager_mutex_);

  auto task = request.task();
  auto learner_id = learner_manager_->GetLearnerId(task.id());
  learner_manager_->UpdateTrainResults(task, learner_id, request.results());
  model_manager_->InsertModel(learner_id, request.model());
  learner_manager_->ScheduleEvaluate({learner_id}, model_manager_->GetModel());

  auto learner_ids = learner_manager_->GetLearnerIds();
  auto to_schedule = scheduler_->ScheduleNext(learner_id, learner_ids.size());

  if (!to_schedule.empty()) {
    model_manager_->UpdateModel(to_schedule, learner_ids);
    learner_manager_->ScheduleTrain(to_schedule, model_manager_->GetModel());
    UpdateTrainParams(to_schedule);
  }

  return absl::OkStatus();
}

void Controller::UpdateTrainParams(
    const std::vector<std::string> &learner_ids) {
  if (global_train_params_.scheduler == "SemiSynchronous") {
    auto global_iteration = scheduler_->GetGlobalIteration();
    if (global_iteration == 2 ||
        global_train_params_.semi_sync_recompute_num_updates) {
      auto semi_sync_lambda = global_train_params_.semi_sync_lambda;
      learner_manager_->UpdateTrainParams(learner_ids, semi_sync_lambda);
    }
  }
}

void Controller::Shutdown() {
  learner_manager_->Shutdown();
  model_manager_->Shutdown();
}
}  // namespace metisfl::controller

File: metisfl/server/core/controller.py

from typing import List
from metisfl.proto import model_pb2, controller_pb2
from metisfl.server.core.learner_manager import LearnerManager
from metisfl.server.core.model_manager import ModelManager
from metisfl.server.scheduling.scheduler import Scheduler
from metisfl.server.selection.scheduled_cardinality import ScheduledCardinality


class Controller:

    learner_manager: LearnerManager = None
    model_manager: ModelManager = None
    selector: ScheduledCardinality = None
    scheduler: Scheduler = None

    def __init__(
        self,
        learner_manager: LearnerManager,
        model_manager: ModelManager,
        selector: ScheduledCardinality,
        scheduler: Scheduler
    ) -> None:
        """Initializes the controller.

        Parameters
        ----------
        learner_manager : LearnerManager
            The learner manager to be used.
        model_manager : ModelManager
            The model manager to be used.
        selector : ScheduledCardinality
            The selector to be used.
        """
        self.learner_manager = learner_manager
        self.model_manager = model_manager
        self.selector = selector
        self.scheduler = scheduler

    def add_learner(self, learner: controller_pb2.Learner) -> str:
        """Adds a learner to the controller.

        Parameters
        ----------
        learner : controller_pb2.Learner
            The learner to be added.
        """
        return self.learner_manager.add_learner(learner=learner)

    def remove_learner(self, learner_id: str) -> None:
        """Removes a learner from the controller.

        Parameters
        ----------
        learner_id : str
            The learner id.
        """
        self.learner_manager.remove_learner(learner_id=learner_id)

    def set_initial_model(self, model: model_pb2.Model) -> None:
        """Sets the initial model for the controller.

        Parameters
        ----------
        model : model_pb2.Model
            The initial model.
        """
        self.model_manager.set_initial_model(model=model)

    def start_training(self) -> None:
        """Starts the training process."""

        learner_ids = self.learner_manager.get_learner_ids()
        for learner_id in learner_ids:
            to_schedule = self.scheduler.schedule(learner_id=learner_id)

            if len(to_schedule) == 0:
                continue

            self.learner_manager.schedule_train(
                learner_ids=to_schedule,
                model=self.model_manager.get_model(),
            )

    def train_done(self, request: controller_pb2.TrainDoneRequest) -> None:
        """Notifies the controller that the training is done.

        Parameters
        ----------
        request : controller_pb2.TrainDoneRequest
            The request containing the learner id and the model.
        """

        task = request.task
        learner_id = self.learner_manager.get_learner_id(task=task)
        model = request.model
        train_results = request.train_result

        self.model_manager.insert_model(
            learner_id=learner_id,
            model=model,
        )

        self.learner_manager.schedule_evaluate(
            learner_ids=[learner_id],
            model=self.model_manager.get_model(),
        )

        self.learner_manager.update_train_result(
            task=task,
            learner_id=learner_id,
            train_results=train_results,
        )

        learner_ids = self.learner_manager.get_learner_ids()
        to_schedule = self.scheduler.schedule(learner_id=learner_id)

        if len(to_schedule) == 0:
            return

        self.model_manager.update_model(
            to_schedule=to_schedule,
            learner_ids=learner_ids,
        )

        self.learner_manager.schedule_train(
            learner_ids=to_schedule,
            model=self.model_manager.get_model(),
        )
